crates/fuel-core/src/graphql_api/worker_service.rs

```diff
@@ -539,6 +544,31 @@ where
     Ok(())
 }
 
+#[derive(Clone, Debug)]
+pub struct BlockHeightSubscriptionHandle {
+    rx_block_height: tokio::sync::watch::Receiver<BlockHeight>,
+}
+
+impl BlockHeightSubscriptionHandle {
+    pub async fn wait_for_block_height(
+        &mut self,
+        height: BlockHeight,
+    ) -> anyhow::Result<()> {
+        loop {
```

Comment:
While it is very simple to implement it with the `tokio::watch`, it is still expensive from an execution perspective.

Let's imagine the current height is 10. The required height for 10k requests is 20.

10 times, each task related to the request will wake up; it is 100k wakeups that check the state(it is read of the RWLock) of the `watch` and wait again.

If you have the list of subscribers for each block height, we can avoid unnecessary wakeups. And it should be very easy to maintain this list via one mutex. You can have `BTreeMap<BlockHeight, Vec<tokio::sync::oneshot::Sender<()>>>`, where you register all listeners, and on each block you just process most left listeners.